File: src/main/java/org/cobbzilla/util/json/JsonNodeComparator.java
package org.cobbzilla.util.json;

import com.fasterxml.jackson.databind.JsonNode;

import java.util.Comparator;

import static org.cobbzilla.util.json.JsonUtil.fromJsonOrDie;

public class JsonNodeComparator implements Comparator<JsonNode> {

    final String path;

    @java.beans.ConstructorProperties({"path"})
    public JsonNodeComparator(String path) {
        this.path = path;
    }

    @Override public int compare(JsonNode n1, JsonNode n2) {
        final JsonNode v1 = fromJsonOrDie(n1, path, JsonNode.class);
        final JsonNode v2 = fromJsonOrDie(n2, path, JsonNode.class);
        if (v1 == null) return 1;
        if (v2 == null) return -1;
        return v1.textValue().compareTo(v2.textValue());
    }
}

File: src/main/java/org/cobbzilla/util/json/JsonNodeComparator.kt
package org.cobbzilla.util.json

import com.fasterxml.jackson.databind.JsonNode

import java.util.Comparator

import org.cobbzilla.util.json.JsonUtil.fromJsonOrDie

class JsonNodeComparator @java.beans.ConstructorProperties("path")
constructor(internal val path: String) : Comparator<JsonNode> {

    override fun compare(n1: JsonNode, n2: JsonNode): Int {
        val v1 = fromJsonOrDie(n1, path, JsonNode::class.java)
        val v2 = fromJsonOrDie(n2, path, JsonNode::class.java)
        if (v1 == null) return 1
        return if (v2 == null) -1 else v1.textValue().compareTo(v2.textValue())
    }
}
